### ISSUE TITLE
Standardize JavaScript and CSS implementations

### ISSUE THREAD

**Issue by richardbporter** [uiowa/uiowa#1977]
Related to #1151, #1402, #4036

We should follow Drupal's lead and start using ES6 syntax for all but the simplest JS. Transpiling is kind of annoying though, and I'd be interested to see if we could just write ES6 syntax without transpiling based on our browser support policy.

Working with project asset directories is confusing as described in #1151. It would be nice if every `assets` directory was ignored and we had a better build process in place to move in all assets. 

Additionally, we seem to have a lot of code duplication between Yarn workspaces. A standardized, top-level `scripts` declaration in the root package.json file would be nice instead of having to copy/paste build scripts everywhere. Another option might a root Gulp file. This may need to wait until #2629 is done.

Also, a lot of our JavaScript is implemented differently. Some is vanilla, some is jQuery, etc. At the very least, all JS should be attached with Drupal behaviors, which is not jQuery, to avoid global variable leak, handle full/partial page reloads, etc.

Also, its important to note that jQuery is loaded on every page because of the dependency specified with BigPipe which renders all page requests. We might want to decide whether or not we want to use jQuery in our scripts. I think it at least needs to be passed to the IIFE and the docs still recommend using `once`: https://www.drupal.org/docs/drupal-apis/javascript-api/javascript-api-overview  Avoiding jQuery doesn't really gain anything other than making a future removal of jQuery a bit easier, whenever that comes.

https://www.drupal.org/node/2815083
https://www.drupal.org/project/drupal/issues/3052002
https://git.drupalcode.org/project/drupal/-/blob/9.2.x/core/modules/big_pipe/big_pipe.libraries.yml#L8

**Comment by GaryRidgway** [comment]
I think we are also at a point where things between UIDS and uids_base are clashing a little bit as well. Generally in UIDS we are using vanilla JS to keep it free of dependencies, and I think this is, by virtue of being the main coding style of UIDS, bleeding into uids_base.

Additionally i completely agree with you about being more careful about polluting the global namespace, and I myself have probably been guilt of not being as careful as i could be in the past.

Honestly though I don't think jquery is much more easy to use or valuable than vanilla ES6 at this point, so i think if vanilla JS has the small benefit of being more long term useful if/when we remove jquery, then we should use that. If I had my choice I would prefer vanilla ES6 with just the jquery "once" attachment function. This uses the recommended attachment function while keeping the core code consistent across both UIDS and uids_base.

**Comment by richardbporter** [comment]
Thanks. I could get on board with that idea. I read this awhile ago and it seems to indicate that passing jQuery to the IIFE and `$.once` may not even be necessary: https://www.previousnext.com.au/blog/removing-jquery-from-your-drupal-theme

https://developers.google.com/web/updates/2016/10/addeventlistener-once

Then whenever jQuery is removed from Drupal core, we'd have an easier time dropping it as our JS would not depend on it.

**Comment by richardbporter** [comment]
Making a note here that we should replace our usage of jquery.cookie with js-cookie as part of this: https://www.drupal.org/node/3104677

**Comment by richardbporter** [comment]
> I'd be interested to see if we could just write ES6 syntax without based on our browser support policy.

Now that https://its.uiowa.edu/news/microsoft-ending-office-365-support-internet-explorer, can we update our browser support policy to drop IE11 support? https://sitenow.uiowa.edu/policies#browser-support

**Comment by joewhitsitt** [comment]
Though this is a big win wasn't there a fair percentage of usage off-campus as well? I am all for this but we can't forget about the recruitment demographic here.

**Comment by richardbporter** [comment]
Non-jQuery `drupal.once` in 9.2.0: https://www.drupal.org/node/3158256

**Comment by richardbporter** [comment]
> We should follow Drupal's lead and start using ES6 syntax for all but the simplest JS. Transpiling is kind of annoying though, and I'd be interested to see if we could just write ES6 syntax without transpiling based on our browser support policy.

I'm not so sure about this anymore. Given that we still need to support IE11 (#4036) coupled with the fact that jQuery is loaded on every page, it seems a heck of a lot easier to just use jQuery. I had a lot of trouble getting transpiled JS to work in #4013 whereas jQuery worked perfectly in IE11.
